flocker/node/agents/cinder.py

```diff
@@ -150,19 +213,51 @@ def list_volumes(self):
 
         See: http://docs.rackspace.com/cbs/api/v1.0/cbs-devguide/content/GET_getVolumesDetail_v1__tenant_id__volumes_detail_volumes.html # noqa
         """
-        volumes = []
-        for cinder_volume in self.volume_manager.list():
+        flocker_volumes = []
+        for cinder_volume in self.cinder_volume_manager.list():
             if _is_cluster_volume(self.cluster_id, cinder_volume):
-                volumes.append(
-                    _blockdevicevolume_from_cinder_volume(cinder_volume)
-                )
-        return volumes
+                flocker_volume = _blockdevicevolume_from_cinder_volume(cinder_volume, self.reverse_host_map)
+                flocker_volumes.append(flocker_volume)
+        return flocker_volumes
+
+    def _get(self, blockdevice_id):
+        for volume in self.list_volumes():
+            if volume.blockdevice_id == blockdevice_id:
+                return volume
+        raise UnknownVolume(blockdevice_id)
 
     def resize_volume(self, blockdevice_id, size):
         pass
 
     def attach_volume(self, blockdevice_id, host):
-        pass
+        """
+        The attaching may have to be done via the nova client :-(
+        See http://www.florentflament.com/blog/openstack-volume-in-use-although-vm-doesnt-exist.html # noqa
+
+        When I attach using the cinder client the volumes become undetachable.
+        """
+        unattached_volume = self._get(blockdevice_id)
+        if unattached_volume.host is not None:
+            raise AlreadyAttachedVolume(blockdevice_id)
+
+        device_path = _next_device()
```

Comment:
You have to supply a device path.
Which means that this API *must* be run on the host to which volumes are being attached.
Which makes me wonder if there's any point having the `host` argument of `IBlockDeviceAPI.attach`?